src/mem/alloc.rs

```diff
@@ -0,0 +1,1063 @@
+/* SPDX-License-Identifier: MIT */
+/*
+ * Copyright (c) 2022 SUSE LLC
+ *
+ * Author: Joerg Roedel <jroedel@suse.de>
+ *
+ * vim: ts=4 sw=4 et
+ */
+
+use crate::cpu::vc_early_make_pages_private;
+use crate::dyn_mem_begin;
+use crate::dyn_mem_end;
+use crate::globals::*;
+use crate::mem::{pgtable_pa_to_va, pgtable_va_to_pa};
+use crate::util::locking::SpinLock;
+use core::alloc::{GlobalAlloc, Layout};
+use core::mem::size_of;
+use core::ptr;
+use x86_64::addr::{align_up, PhysAddr, VirtAddr};
+use x86_64::structures::paging::frame::PhysFrame;
+
+struct PageStorageType(u64);
+
+// Support only order 0 allocations for now
+pub const MAX_ORDER: usize = 6;
+
+impl PageStorageType {
+    pub const fn new(t: u64) -> Self {
+        PageStorageType(t)
+    }
+
+    fn encode_order(&self, order: usize) -> PageStorageType {
+        PageStorageType(self.0 | ((order as u64) & PAGE_ORDER_MASK) << PAGE_TYPE_SHIFT)
+    }
+
+    fn encode_next(&self, next_page: usize) -> PageStorageType {
+        PageStorageType(self.0 | (next_page as u64) << PAGE_FREE_NEXT_SHIFT)
+    }
+
+    fn encode_slab(slab: VirtAddr) -> Self {
+        PageStorageType(PAGE_TYPE_SLABPAGE | slab.as_u64() & PAGE_TYPE_SLABPAGE_MASK)
+    }
+}
+
+const PAGE_TYPE_SHIFT: u64 = 4;
```

Comment:
Constant definitions should include a comment with their value. This is for website documentation purposes (see CONTRIBUTING.md). For example, in this case it would be:

```rust
/// 4
const PAGE_TYPE_SHIFT: u64 = 4;

```